src/worker/triggers/keyword.ts

```diff
@@ -71,40 +74,45 @@ const Defaults = {
  *     in the range [0, 1]
  */
 export default class KeywordRecognizer implements SpeechProcessor {
-  private models?: CommandModels
+  private config: Required<KeywordTriggerConfig>
+  private models: CommandModels
   private hopSamples: number
-  private melWidth: number
-  private sampleWindow: RingBuffer<number> = new RingBuffer<number>(0)
-  private encodeWindow: RingBuffer<Tensor> = new RingBuffer<Tensor>(0)
-  private encodeState: Tensor = tf.zeros([1])
-  private frameWindow: RingBuffer<Tensor> = new RingBuffer<Tensor>(0)
-  private keywordClasses: string[]
-  private threshold: number
+  private sampleWindow = new RingBuffer<number>(0)
+  private encodeWindow = new RingBuffer<Tensor>(0)
+  private encodeState = self.tf.zeros([1])
+  private frameWindow = new RingBuffer<Tensor>(0)
   private vadActive = false
 
-  constructor(config: SpeechConfig) {
-    // number of samples to slide the window for each filter run
-    this.hopSamples = (config.hopLength || Defaults.hopLength) * (config.sampleRate / 1000)
-    this.melWidth = config.melWidth || Defaults.melWidth
-    // if (classes.length != expected) {
+  static async create(config: SpeechConfig) {
+    if (!config.baseKeywordUrl) {
+      throw new Error('keyword URL required')
+    }
     if (!config.keywordClasses || config.keywordClasses.length == 0) {
       throw new Error('keyword classes required')
-      // return new Error(`invalid number of keyword classes: ${classes.length} != ${expected}`)
     }
-    this.keywordClasses = config.keywordClasses
-    this.threshold = config.keywordThreshold || Defaults.threshold
+    config.sampleRate = Number(config.sampleRate)
+    if (isNaN(config.sampleRate)) {
```

Comment:
Shouldn't this be impossible, given that we always have a fallback default?

Reply:
It would be impossible without a type error in typescript, but I've added a similar guard in response to a recent issue that passed a string. Checking `isNaN` ensures it's a valid number.